File: CrackingTheCodingInterview/TechnicalQuestions/ReshapingMatrix.js
var matrixReshape = function(mat, r, c) {
    if (mat.length == 0 || c * r != (mat.length * mat[0].length)) return mat;
    let res = new Array(r).fill(0).map(() => new Array(c).fill(0))
    //otherwise, return original matrix
    console.log(res)
    if (r ==  mat.length && c == mat[0].length) return mat;
    let row = 0, col = 0;
    for (let i = 0; i < mat.length; i++) {
        for (let j = 0; j < mat[0].length; j++) {
            res[row][col] = mat[i][j];
            if (col < c-1) {
                col++;
            } else if (col == c-1) {
                col = 0;
                row++;
            }
        }
    }
    return res;
};

File: CrackingTheCodingInterview/TechnicalQuestions/ReshapingMatrix.java
package CrackingTheCodingInterview.TechnicalQuestions;

import java.util.*;

public class ReshapingMatrix {
    public static void main(String[] args) {
        
    }

    public int[][] matrixReshape(int[][] mat, int r, int c) {
        int[][] res = new int[r][c];
        if (mat.length == 0 || r * c != (mat[0].length * mat.length)) {
            return mat;
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat.length; j++){
                queue.add(mat[i][j]);
            }
        }
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                res[i][j] = queue.remove();
            }
        }
        return res;
    }
}